### PR TITLE
8319873: Add windows implementation for jcmd System.map and System.dump_map

### PR DESCRIPTION
**Pull request by stooke**
This is a port of [JDK-8318636](https://github.com/openjdk/jdk/pull/16301) to Windows.

System.map and System.dump_map are implemented using the Windows API and provide roughly the same information in the same format.  Most of the heavy lifting was implemented by @tstuefe in #16301 - this PR adds the Windows implementation and enables the common code for Windows 64 bit.
 
[Sample output (with NMT enabled)](https://github.com/user-attachments/files/16982072/sample_windows_map.txt)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319873](https://bugs.openjdk.org/browse/JDK-8319873): Add windows implementation for jcmd System.map and System.dump_map (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Sonia Zaldana Calles](https://openjdk.org/census#szaldana) (@SoniaZaldana - Committer) Review applies to [c87345ad](https://git.openjdk.org/jdk/pull/20597/files/c87345adddcc80480b3db4890ab90dfb7b9563dc)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)

### Contributors
 * Simon Tooke `<stooke@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20597/head:pull/20597` \
`$ git checkout pull/20597`

Update a local copy of the PR: \
`$ git checkout pull/20597` \
`$ git pull https://git.openjdk.org/jdk.git pull/20597/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20597`

View PR using the GUI difftool: \
`$ git pr show -t 20597`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20597.diff">https://git.openjdk.org/jdk/pull/20597.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20597#issuecomment-2297057958)

### CODE REVIEW

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (tstuefe) [inline src/hotspot/os/windows/memMapPrinter_windows.cpp:162]: The "danger" here is that both of these functions may bloat in the future and print a lot more than today; and that we may want to use this printout in places where we also print out OS info and memory info, and therefore get duplicated printout (e.g. hs-err printing or jcmd VM.info). 

But I would print out the number of mappings you counted, and the total size of committed regions.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (tstuefe) [inline src/hotspot/share/services/diagnosticCommand.cpp:1207]: I wince a bit a the new ifdefs. Possibly for a separate RFE, it would be nice to have a function os::realpath() that uses realpath(2) on POSIX platforms, _fullpath on Windows.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (tstuefe) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp:98]: But (here and in other places) raw-printing the unknown constant may not be the best way. Chances are this code is executed rarely and most people will just be perplexed at the weird printouts.

We could assert here: After all, we expect to handle all possible values; if we see an unknown constant, then we either have a bug in the code and what we read here is no protection value, or Microsoft added new constants. In both cases, a code fix is needed. Asserts would fire up in automatic regression tests.

The alternative is to print an '?'. And possibly scan for that in regression tests too.

 I leave this decision up to you.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (tstuefe) [inline src/hotspot/os/windows/memMapPrinter_windows.cpp:246]: I would add a failsafe breakout in case we made a thinking error somewhere or the address space is insanely fragmented. Maybe break out after a million regions? Or, after a certain time (e.g. 10 seconds). I also would assert for RegionSize > 0

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (tstuefe) [inline src/hotspot/os/windows/memMapPrinter_windows.cpp:243]: Just to be sure, I would zero out mInfo.

**Review by tstuefe — COMMENTED** [review]
Hi Simon,

this is nice, thank you!

Mostly nits.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: done.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: done and done.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (stooke) [reply src/hotspot/share/services/diagnosticCommand.cpp:1207]: I agree, and was surprised this wasn't originally implemented.  I'll prepare a second PR.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I've done this and revised the legend, too.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I've rewritten this code to use stringStream instead.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: dont - and I noticed tabs were inconsistent too.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I honestly have no clue how iostreams crept into there.  Gone!  and C header now used.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: Done.

**Review by tstuefe — APPROVED** [review]
Good! I like the new logic in get_protect_string.

**Comment by tstuefe** [comment]
Needs a second reviewer. @jdksjolen maybe?

**Comment by plummercj** [comment]
I think `test/jdk/sun/tools/jcmd/TestJcmdPIDSubstitution.java` should be updated to allow System.dump_map on Windows, not just linux.

https://github.com/openjdk/jdk/commit/6733b89de1ed9ebcc1e8a8271970048ad0f0aac3

**Comment by tstuefe** [comment]
> I think `test/jdk/sun/tools/jcmd/TestJcmdPIDSubstitution.java` should be updated to allow System.dump_map on Windows, not just linux.
> 
> [6733b89](https://github.com/openjdk/jdk/commit/6733b89de1ed9ebcc1e8a8271970048ad0f0aac3)

Make sense. Should work out of the box. @stooke this is a test Sonia added some weeks ago, see https://bugs.openjdk.org/browse/JDK-8334492

**Comment by stooke** [comment]
> > I think `test/jdk/sun/tools/jcmd/TestJcmdPIDSubstitution.java` should be updated to allow System.dump_map on Windows, not just linux.
> > [6733b89](https://github.com/openjdk/jdk/commit/6733b89de1ed9ebcc1e8a8271970048ad0f0aac3)
> 
> Make sense. Should work out of the box. @stooke this is a test Sonia added some weeks ago, see https://bugs.openjdk.org/browse/JDK-8334492

I have enabled this test for Windows.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (stooke) [reply src/hotspot/share/services/diagnosticCommand.cpp:1207]: I have filed issue [#8338851](https://bugs.openjdk.org/browse/JDK-8338851) "Hoist os::Posix::realpath() to os::realpath() and implement on Windows" and will submit a PR shortly.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I have rolled my own "print_os_info" from the utility functions, but have left print_memory_info in for now - if you want, I will remove it entirely, but I think the information provided is valuable to readers of this report.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I now handle less cases but will assert() if unimplemented bits are encountered.  I have currently only implemented this for protection bits, but I'm still thinking what's best and intent to handle unknown cases identically for the other two functions.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (stooke) [reply src/hotspot/share/services/diagnosticCommand.cpp:1207]: I have created PR #20683 to address this.

**Review by tstuefe — APPROVED** [review]
Still good

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I've also printed the region count.

**Review by tstuefe — APPROVED** [review]
Still good

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (SoniaZaldana) [inline src/hotspot/os/windows/memMapPrinter_windows.cpp:98]: Small nit - indentation.

File: src/hotspot/share/services/diagnosticCommand.hpp
Comment (SoniaZaldana) [inline src/hotspot/share/services/diagnosticCommand.hpp:984]: We also need to update the description for the diagnostic commands. Currently, both `System.dump_map` and `System.map` state they're enabled for linux only.

Comment (stooke) [reply src/hotspot/share/services/diagnosticCommand.hpp:984]: A good catch - I'd already found this while preparing a macOS implementation.
Fixed.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: Fixed.

**Comment by stooke** [comment]
@SoniaZaldana thank you for looking at my PR.  I've fixed the issues you've found.  

As this "reopens" the PR, I also took the opportunity to fix an issue when no filename is given to the '-F' parameter in System.dump_map (i.e. when "jcmd 999 System.dump_map -F" is issued, for example).  Previously, the Windows implementation would terminate the VM (an error in fileStream::fileStream(const char*) presumably) and crash jcmd with a java stack trace.  (this should be addressed in a separate PR).

Now an error message is sent to the jcmd user: 
`C:\work>.\jdk\build\fastdebug\jdk\bin\jcmd TT System.dump_map -F=`
`13008:`
`filename is empty or not specified.  No file written`

I'm sorry to ask, @tstuefe , but this impacts the Linux implementation too, so you might want to look at this new behaviour in SystemDumpMapDCmd::execute() (diagnosticComand.cpp:1196)

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (tstuefe) [inline src/hotspot/share/services/diagnosticCommand.cpp:1196]: Can you please change this to `(*name) == '\0'` or `name[0] == '\0'` ?

**Review by tstuefe — COMMENTED** [review]
No problem. Small nit, otherwise still fine.

**Comment by SoniaZaldana** [comment]
Hi Simon, 

Thanks! Looks good. 

Just had a question about the empty filename issue. I don’t think there is a case where a filename parameter should accept an empty string (but I might be wrong). 

If that’s the case, I wonder if it would be better to just deal with all empty filename parameter errors in the dcmd argument parser. 

For example, in [diagnosticArgument.cpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/services/diagnosticArgument.cpp#L186
).  Perhaps, we could have a check like this: 

```
if (strcmp(type(), "FILE") == 0) {
     if (str == NULL || *str == 0) {
       stringStream error_msg;
       error_msg.print("Filename is empty or not specified. %s", str);
       THROW_MSG(vmSymbols::java_lang_IllegalArgumentException(), error_msg.base());
     }
     _value = REALLOC_C_HEAP_ARRAY(char, _value, JVM_MAXPATHLEN, mtInternal);
     ...
   }
```

I haven't tested this but this would probably be better handled in a separate RFE though.

**Review by tstuefe — APPROVED** [review]
ok

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp:98]: I was thinking about this - my original intent was to be fatal only on a debug build, not release.  I have added a message in the output, and changed these to warnings (for both release and debug)

Comment (tstuefe) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp:98]: I would prefer an assert here. It will crash in debug builds, which is fine, since if it crashes it is either a bug or an hitherto unknown state. In any case something that should be fixed.

Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp:98]: I believe we can all agree that fatal() was too harsh.
Is there a macro that is the equivalent of "assert(false, ....)"?  
my current unchecked in code has 'warning()' which will log the problem.  I am happy to replace this with assert().
How do you feel about a message in the output even in release builds?

**Comment by SoniaZaldana** [comment]
> 
> I agree, and will probably do this next week. If it looks good to you, could you please mark as reviewed? You're on the list of reviewers.

FYI, I filed [JDK-8339970](https://bugs.openjdk.org/browse/JDK-8339970) to track this.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (dholmes-ora) [inline src/hotspot/share/services/diagnosticCommand.cpp:1196]:
```suggestion
  if (name == nullptr
```

**Comment by stooke** [comment]
Hello @kevinjwalls , and thank you for your review!  I have attempted to address your concerns with my use of 'fatal()' by replacing with a message in the returned output and an assert().

I have also adjusted the spacing of the offset field in the output, but kept it for parity with the Linux version.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (SoniaZaldana) [inline src/hotspot/os/windows/memMapPrinter_windows.cpp:212]: I think the extended "===" in the header was okay :-)

**Review by SoniaZaldana — APPROVED** [review]
Looks good to me.

**Comment by kevinjwalls** [comment]
> concerns with my use of 'fatal()' by replacing with a message in the returned output and an assert().
> 
> I have also adjusted the spacing of the offset field in the output, but kept it for parity with the Linux version.

Thanks @stooke !

Thanks for making it non-fatal.
Printing a message on an unknown field will break the format?  I don't want to cause a great disturbance on this as it should "never happen" so this is fine, or doing nothing and having the assert for debug builds.

Re: offset column: I don't see an offset column in Linux, I see an rss column?

**Comment by stooke** [comment]
@kevinjwalls 
> Re: offset column: I don't see an offset column in Linux, I see an rss column?
You're right, I missed that.  As far as I can tell, that information isn't available at that level for a memory segment.
Are you okay with leaving the offset in?  One bit of information lost without offset which memory block points to AllocationBase.

File: src/hotspot/os/windows/memMapPrinter_windows.cpp
Comment (stooke) [reply src/hotspot/os/windows/memMapPrinter_windows.cpp]: I stuck a couple extra '===' in, just for you.  I do think it looks a bit better.

**Comment by kevinjwalls** [comment]
Sorry if I'm holding this up.  The reasoning did get confusing to me.

We are not always good at documenting these commands.  But at some point, somebody among us will hopefully document them, or explain it to somebody who will.  There will need to be explanation of how the outputs differ on Linux and Windows, and how the information can be used.

RSS is on Linux only... No problem, that has a well-understood meaning on *nix systems.

offset?  It's the same as the size of the previous line of output.  Or the sum of sizes if there are multiple lines as part of the same allocation.  It's Windows only.  It's not likely this is more useful on Windows than Linux.  Is it considered missing on Linux, or just not needed?

**Comment by kevinjwalls** [comment]
Thanks Thomas - 
It's understood what the info is, main question was when is it useful, and do we need to differ from Linux.

Looking at 8318636: https://github.com/openjdk/jdk/pull/16301
the Linux example linked in the introduction there shows an offset column.

Maybe Simon you're taking inspiration from that original, 
and there was an extension to the feature just over a month ago in
8322475: https://github.com/openjdk/jdk/pull/17158

...where the offset column was taken out.
8322475 also separated "info" and "file" into their own columns.  Only CDS and one of the hsperfdata lines use both columns that I notice, but there may be others.
Here in Windows we're going with one column for both, unlike the updated Linux format.

This PR is implementing "classic" System.map, not the "modern 2024" version? 8-)

It's a good feature to have, I am just being annoyingly clear about when and why we differ between platforms.

**Comment by stooke** [comment]
@kevinjwalls I believe we all expect that information will be different between platforms, and some information is more important on one platform than another.  For example I have a rough implementation of a macOS System.map and it will differ again, although as the same author, it will have some similarities (e.g. merged vminfo/file field) to this implementation.  I agree the offset column isn't vital, but the start of allocation may be (at some point, to someone), and, as @tstuefe commented, there is no equivalent on Linux.

**Comment by stooke** [comment]
@tstuefe I have updated the legend, thanks for the note.

**Review by kevinjwalls — APPROVED** [review]
Thanks Simon, I think I should click to register happiness at this point. 8-)

**Comment by tstuefe** [comment]
>I agree the offset column isn't vital, but the start of allocation may be (at some point, to someone), and, as @tstuefe commented, there is no equivalent on Linux.

The offset is good to know since it shows clearly what original VirtualAlloc allocation a committed section belongs to. E.g. all committed/uncommitted sections of the heap should refer in their offsets to the java heap.

**Comment by tstuefe** [comment]
> Thanks Thomas - It's understood what the info is, main question was when is it useful, and do we need to differ from Linux.
> 
> Looking at 8318636: #16301 the Linux example linked in the introduction there shows an offset column.
> 
> Maybe Simon you're taking inspiration from that original, and there was an extension to the feature just over a month ago in 8322475: #17158
> 
> ...where the offset column was taken out. 8322475 also separated "info" and "file" into their own columns. Only CDS and one of the hsperfdata lines use both columns that I notice, but there may be others. Here in Windows we're going with one column for both, unlike the updated Linux format.
> 
> This PR is implementing "classic" System.map, not the "modern 2024" version? 8-)
> 

"offset" in the Linux version was the file offset of mapped files. We have only very few mapped files, most mappings are anonymous, and the offset is usually 0. I took it out to save horizontal space.

**Review by tstuefe — APPROVED** [review]
Still good. Ship it!

**Comment by dholmes-ora** [comment]
@stooke this is causing a build failure in our CI with a slowdebug build:
```
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(126): error C2504: 'os::win32': base class undefined
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(130): error C2027: use of undefined type 'os::win32'
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\share\runtime/os.hpp(1015): note: see declaration of 'os::win32'
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(130): error C3861: 'print_windows_version': identifier not found
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(131): error C2027: use of undefined type 'os::win32'
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\share\runtime/os.hpp(1015): note: see declaration of 'os::win32'
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(131): error C3861: 'print_uptime_info': identifier not found
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(132): error C2653: 'VM_Version': is not a class or namespace name
[2024-09-18T11:35:44,152Z] c:\sb\prod\1726658372\workspace\open\src\hotspot\os\windows\memMapPrinter_windows.cpp(132): error C3861: 'print_platform_virtualization_info': identifier not found
```
I can't file a bug at present but please attend to this ASAP.

**Comment by stefank** [comment]
The build failure happens when the build is configured with '--disable-precompiled-headers'. FWIW, I recommend that everyone who make changes to HotSpot make sure to use that configure flag. I wish that it was the default build configuration, maybe it is time to change this.

**Comment by jaikiran** [comment]
> I can't file a bug at present but please attend to this ASAP.

I have filed https://bugs.openjdk.org/browse/JDK-8340368 to track this.

**Comment by kevinjwalls** [comment]
We have some test failures also, logged https://bugs.openjdk.org/browse/JDK-8340391

**Comment by stefank** [comment]
There's also a new failure here:
https://bugs.openjdk.org/browse/JDK-8340401

**Pull request closed without merging**
